src/server/bg_services/objects_reclaimer.js

```diff
@@ -0,0 +1,64 @@
+/* Copyright (C) 2016 NooBaa */
+'use strict';
+
+const config = require('../../../config');
+const dbg = require('../../util/debug_module')(__filename);
+const MDStore = require('../object_services/md_store').MDStore;
+const system_store = require('../system_services/system_store').get_instance();
+const system_utils = require('../utils/system_utils');
+const map_deleter = require('../object_services/map_deleter');
+const P = require('../../util/promise');
+
+class ObjectsReclaimer {
+
+    constructor({ name, client }) {
+        this.name = name;
+        this.client = client;
+    }
+
+    async run_batch() {
+        if (!this._can_run()) return;
+
+        const unreclaimed_objects = await MDStore.instance().find_unreclaimed_objects(config.OBJECT_RECLAIMER_BATCH_SIZE);
+        if (!unreclaimed_objects || !unreclaimed_objects.length) {
+            dbg.log0('no objects in "unreclaimed" state. nothing to do');
+            return config.OBJECT_RECLAIMER_EMPTY_DELAY;
+        }
+
+        let has_errors = false;
+        dbg.log0('object_reclaimer: starting batch work on objects: ', unreclaimed_objects.map(o => o.key).join(', '));
+        const reclaimed_objects_ids = [];
+        await P.all(unreclaimed_objects.map(async obj => {
+            try {
+                await map_deleter.delete_object_mappings(obj, true);
+                reclaimed_objects_ids.push(obj._id);
+            } catch (err) {
+                dbg.error(`got error when trying to delete object ${obj.key} mappings :`, err);
+                has_errors = true;
+            }
+        }));
+        await MDStore.instance().update_objects_by_ids(reclaimed_objects_ids, { reclaimed: new Date() });
+
+        if (has_errors) {
+            return config.OBJECT_RECLAIMER_ERROR_DELAY;
+        }
+        return config.OBJECT_RECLAIMER_BATCH_DELAY;
+
+    }
+
+    _can_run() {
+        if (!system_store.is_finished_initial_load) {
+            dbg.log0('MirrorWriter: system_store did not finish initial load');
```

Comment:
first rule in school - when you copy, don't copy the name..

Reply:
:sweat_smile: